sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala

```diff
@@ -396,8 +396,9 @@ case class AlterTableChangeColumnCommand(
     val newDataSchema = table.dataSchema.fields.map { field =>
       if (field.name == originColumn.name) {
         // Create a new column from the origin column with the new comment.
+        val newField = field.copy(dataType = newColumn.dataType)
```

Comment:
Can you make it more explicit that we effectively allow the data type to change now but we only allow type changes that differ by collation?
E.g. by splitting `columnEqual` into checking that names match on one side and that the type change is supported on the other + add a ` withNewType` method.

Reply:
Clarified in comment, split the `columnEqual` to two functions and added `withNewType` in `StructField`, please check.